### ISSUE TITLE
npm ERR! Failed at the canvas@1.1.6 install script 'node-gyp rebuild'

### ISSUE THREAD

**Issue by costas90** [Automattic/node-canvas#742]
I get the following errors, related to canvas, when trying to install fabricJS.

I am on windows, I followed the installation instructions and installed canvas via npm without a problem. The problem is when trying to install fabricJS.

```
$ npm install fabric --save
npm WARN package.json jsfabric_test@1.0.0 No repository field.
npm WARN package.json jsfabric_test@1.0.0 No README data
WARN engine fabric@1.6.0-rc.1: wanted: {"node":">=0.4.0 && <1.0.0"} (current: {"node":"4.3.2","npm":"2.14.12"})

> canvas@1.1.6 install D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas
> node-gyp rebuild


D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node  rebuild )
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
  Canvas.cc
  CanvasGradient.cc
  CanvasPattern.cc
  CanvasRenderingContext2d.cc
  color.cc
  Image.cc
  ImageData.cc
  init.cc
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\CanvasGradient.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): note: command-line arguments:  see previous definition of 'snprintf' (compiling source file ..\src\CanvasGradient.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1927): fatal error C1189: #error:  Macro definition of snprintf conflicts with Standard Library function declaration (compiling source file ..\src\CanvasGradient.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\CanvasPattern.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): note: command-line arguments:  see previous definition of 'snprintf' (compiling source file ..\src\CanvasPattern.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1927): fatal error C1189: #error:  Macro definition of snprintf conflicts with Standard Library function declaration (compiling source file ..\src\CanvasPattern.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
  PixelArray.cc
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\Canvas.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): note: command-line arguments:  see previous definition of 'snprintf' (compiling source file ..\src\Canvas.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1927): fatal error C1189: #error:  Macro definition of snprintf conflicts with Standard Library function declaration (compiling source file ..\src\Canvas.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\ImageData.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): note: command-line arguments:  see previous definition of 'snprintf' (compiling source file ..\src\ImageData.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1927): fatal error C1189: #error:  Macro definition of snprintf conflicts with Standard Library function declaration (compiling source file ..\src\ImageData.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\color.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): note: command-line arguments:  see previous definition of 'snprintf' (compiling source file ..\src\color.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1927): fatal error C1189: #error:  Macro definition of snprintf conflicts with Standard Library function declaration (compiling source file ..\src\color.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\Image.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): note: command-line arguments:  see previous definition of 'snprintf' (compiling source file ..\src\Image.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1927): fatal error C1189: #error:  Macro definition of snprintf conflicts with Standard Library function declaration (compiling source file ..\src\Image.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\init.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): note: command-line arguments:  see previous definition of 'snprintf' (compiling source file ..\src\init.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1927): fatal error C1189: #error:  Macro definition of snprintf conflicts with Standard Library function declaration (compiling source file ..\src\init.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\PixelArray.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): note: command-line arguments:  see previous definition of 'snprintf' (compiling source file ..\src\PixelArray.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1927): fatal error C1189: #error:  Macro definition of snprintf conflicts with Standard Library function declaration (compiling source file ..\src\PixelArray.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): warning C4005: 'snprintf': macro redefinition (compiling source file ..\src\CanvasRenderingContext2d.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1925): note: command-line arguments:  see previous definition of 'snprintf' (compiling source file ..\src\CanvasRenderingContext2d.cc)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.10240.0\ucrt\stdio.h(1927): fatal error C1189: #error:  Macro definition of snprintf conflicts with Standard Library function declaration (compiling source file ..\src\CanvasRenderingContext2d.cc) [D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas\build\canvas.vcxproj]
gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\MSBuild\14.0\bin\msbuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:270:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Windows_NT 6.3.9600
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\canvas
gyp ERR! node -v v4.3.2
gyp ERR! node-gyp -v v3.0.3
gyp ERR! not ok

> contextify@0.1.15 install D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\jsdom\node_modules\contextify
> node-gyp rebuild


D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\jsdom\node_modules\contextify>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node  rebuild )
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
  contextify.cc
  win_delay_load_hook.c
     Creating library D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\jsdom\node_modules\contextify\build\Release\contextify.lib and object D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\jsdom\node_modules\contextify\build\Release\contextify.exp
  Generating code
  Finished generating code
  contextify.vcxproj -> D:\Drive\Personal\Sandbox\jsfabric\node_modules\fabric\node_modules\jsdom\node_modules\contextify\build\Release\\contextify.node
npm ERR! Windows_NT 6.3.9600
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "fabric" "--save"
npm ERR! node v4.3.2
npm ERR! npm  v2.14.12
npm ERR! code ELIFECYCLE

npm ERR! canvas@1.1.6 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the canvas@1.1.6 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the canvas package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls canvas
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     D:\Drive\Personal\Sandbox\jsfabric\npm-debug.log
```
`

I looked at many related threads but I cant understand what is the proper solution to this, I am a beginner programmer. This thread is identical to my problem >>> https://github.com/Automattic/node-canvas/issues/722

Attached is the npm debug log

[npm-debug.txt](https://github.com/Automattic/node-canvas/files/185770/npm-debug.txt)

**Comment by LinusU** [comment]
This have been fixed in a later version, please try with the latest version which at the moment is 1.3.12

**Comment by costas90** [comment]
But I am using 1.3.12

![capture](https://cloud.githubusercontent.com/assets/18027867/13982627/d42c5f24-f0f3-11e5-98ba-2ff226349bf4.PNG)

In what directory should I run node-gyp rebuild. I think I didn't run this command. Is it related?

**Comment by LinusU** [comment]
Hmm, you are installing `fabric`, maybe it depends on an older version?

**Comment by LinusU** [comment]
You can try `npm install canvas@1.3.12`

**Comment by LinusU** [comment]
You shouldn't have to run `node-gyp` manually

**Comment by costas90** [comment]
From the repo in https://github.com/kangax/fabric.js/blob/master/package.json, it does depend on an older version:

`  "dependencies": {
    "canvas": "1.2.x",
    "jsdom": "1.1.x",
    "xmldom": "0.1.x"
  },`

I added the package manually just like it was suggested in https://github.com/Automattic/node-canvas/issues/722#issuecomment-185619583, but I am not sure how to use npm shrinkwrap. Do  I have to use it?

I test fabric and it works.

**Comment by LinusU** [comment]
I'm not sure actually, maybe the easiest would be if fabric could bump their dependency, have you opened an issue with them?

**Comment by costas90** [comment]
I havent but I will. Thanks

**Comment by AbdurRehman91** [comment]
hi i am installing farbic.js 1.6.0 and i am getting same error.help needed

**Comment by costas90** [comment]
@AbdurRehman91 What I did, I added the fabricJS files manually on the modules directory. Then I  changed the fabricJS dependency for canvas from 1.2.x to 1.3.12 in package.json. Then I changed directory to the fabricJS directory I run npm install to update the canvas dependencies.

I hope this makes sense. Not sure if there is a faster way to solve this.

**Comment by yamiko-ninja** [comment]
1.6.2 has the same error

**Comment by michaeljfalk** [comment]
Confirming that I am also getting the same error as @yamiko-ninja on version 1.6.2

**Comment by anthonyettinger** [comment]
getting this too on 1.6.2

**Comment by sergomet** [comment]
same on 1.6.2

**Comment by chearon** [comment]
I investigated OP's error message a bit, something is redefining `snprintf` but the compiler isn't telling us what is doing it specifically. I think the error message [was added for VS2015](http://stackoverflow.com/a/27754829/487014).

Actually I'm seeing `C:/GTK/include/libintl.h` define `snprintf` on line 375, if someone wants to try deleting that line and then redoing the `npm install`, that might work. I don't get the error, though...

Also, #722 is not related to this
